addons/protected_fields/models.py

```diff
@@ -52,12 +52,13 @@ def _check_field_permissions(self, vals, permitted_vals=None):
     def create(self, vals):
         # default values are permitted upon creation
         defaults = self.default_get(vals)
-        self._check_field_permissions(vals, permitted_vals=defaults)
         record = super(ProtectedFieldsMixin, self).create(vals)
+        record._check_field_permissions(vals, permitted_vals=defaults)
         return record
 
     @api.multi
     def write(self, vals):
         self._check_field_permissions(vals)
-        record = super(ProtectedFieldsMixin, self).write(vals)
-        return record
+        success = super(ProtectedFieldsMixin, self).write(vals)
+        self._check_field_permissions(vals)
```

Comment:
Is it really scary? It's all performed in the same Postgresql transaction, which is automatically rolled back if the permission test fails. Also note that the test is performed twice - both before and after the write (the first one checks if the user has permissions to change the fields at all, the second one ensures the object wasn't changed beyond those permissions).